File: firmware/src/protocol.c
#include "protocol.h"
#include <stdio.h>
#include <string.h>

Command protocol_parse_command(const char *line) {
    Command cmd;
    memset(&cmd, 0, sizeof(cmd));
    cmd.type = CMD_INVALID;

    if (line == NULL || line[0] == '\0') {
        cmd.type = CMD_NONE;
        return cmd;
    }

    switch (line[0]) {
        case 'K': {
            float kp, ki, kd;
            if (sscanf(line, "K,%f,%f,%f", &kp, &ki, &kd) == 3) {
                cmd.type = CMD_SET_GAINS;
                cmd.kp = kp;
                cmd.ki = ki;
                cmd.kd = kd;
            }
            break;
        }
        case 'S': {
            float setpoint;
            if (sscanf(line, "S,%f", &setpoint) == 1) {
                cmd.type = CMD_SET_SETPOINT;
                cmd.setpoint = setpoint;
            }
            break;
        }
        case 'M': {
            int mode;
            if (sscanf(line, "M,%d", &mode) == 1 &&
                (mode == MODE_POSITION || mode == MODE_SPEED)) {
                cmd.type = CMD_SET_MODE;
                cmd.mode = mode;
            }
            break;
        }
        case 'E': {
            int enable;
            if (sscanf(line, "E,%d", &enable) == 1) {
                cmd.type = CMD_ENABLE;
                cmd.enable = enable ? 1 : 0;
            }
            break;
        }
        case 'R': {
            cmd.type = CMD_RESET;
            break;
        }
        default:
            break;
    }

    return cmd;
}

int protocol_format_telemetry(char *buf, size_t bufsize, const Telemetry *t) {
    int n = snprintf(buf, bufsize, "T,%lu,%d,%.4f,%.4f,%.4f,%.4f,%.4f\n",
                      (unsigned long)t->t_ms, t->mode, t->setpoint,
                      t->position, t->speed, t->error, t->output);
    if (n < 0 || (size_t)n >= bufsize) {
        return -1;
    }
    return n;
}

File: dashboard/protocol.py
"""Python mirror of firmware/include/protocol.h.

Speaks the identical line-based ASCII protocol used by both the STM32
firmware (over UART) and the native simulation binary (over stdio), so
this module -- and everything built on it -- doesn't care which one
it's talking to.
"""
from dataclasses import dataclass

MODE_POSITION = 0
MODE_SPEED = 1


@dataclass
class Telemetry:
    t_ms: int
    mode: int
    setpoint: float
    position: float
    speed: float
    error: float
    output: float


def parse_telemetry(line: str):
    """Parses one 'T,...' telemetry line. Returns None if malformed."""
    line = line.strip()
    if not line.startswith("T,"):
        return None
    parts = line.split(",")
    if len(parts) != 8:
        return None
    try:
        return Telemetry(
            t_ms=int(parts[1]),
            mode=int(parts[2]),
            setpoint=float(parts[3]),
            position=float(parts[4]),
            speed=float(parts[5]),
            error=float(parts[6]),
            output=float(parts[7]),
        )
    except ValueError:
        return None


def cmd_set_gains(kp: float, ki: float, kd: float) -> str:
    return f"K,{kp},{ki},{kd}\n"


def cmd_set_setpoint(value: float) -> str:
    return f"S,{value}\n"


def cmd_set_mode(mode: int) -> str:
    return f"M,{mode}\n"


def cmd_enable(enable: bool) -> str:
    return f"E,{1 if enable else 0}\n"


def cmd_reset() -> str:
    return "R\n"
